### PR TITLE
A few frozen string and symbol optimizations

### PR DESCRIPTION
**Pull request by headius**
This cleans up some object flags in order to start flagging fstrings as such and use that to cache any symbol they might intern.

### CODE REVIEW

**Comment by headius** [comment]
This seemed like a great idea until I realized you can't add new internal variables to a frozen object (or even update them if they have been allocated already).

Still feels like there's some value in having fstrings cache common products like symbols and regexp.

**Review by enebo — COMMENTED** [review]
We probably need a comment about how the CR values cannot change.

File: core/src/main/java/org/jruby/util/StringSupport.java [hunk 1]
```diff
@@ -72,8 +72,8 @@ public final class StringSupport {
     public static final int CR_UNKNOWN   = 0;
 
     // We hardcode these so they can be used in a switch below. The assert verifies they match FlagRegistry's value.
-    public static final int CR_7BIT      = 16;
-    public static final int CR_VALID     = 32;
+    public static final int CR_7BIT      = 8;
```
[the remark this answers is not in the document]
Comment (headius) [reply core/src/main/java/org/jruby/util/StringSupport.java]: I corrected these values and added comments and harder errors to make sure they don't change.

**Comment by headius** [comment]
The original FString idea has been implemented with a subclass of RubyString that aggregates lazy fields for symbol, integer, and float. Performance is greatly improved for cases that request those conversions from FStrings.

**Pull request merged**